_posts/2021-08-10-newpipe-0.21.9-released.md

```diff
@@ -0,0 +1,148 @@
+
+- You can now add the same stream to a queue multiple times as you prefer, thanks to [@Stypox](https://github.com/Stypox). On earlier versions, this would result in endless buffering. So this is a bug fix and a feature rolled into one!
+
+- Sometimes there could be an old invalid download in the Downloads page which caused a crash on starting new downloads, so [@Stypox](https://github.com/Stypox) made the app handle that.
+
+- While he was at it, [@Stypox](https://github.com/Stypox) also fixed a different error, an exception that occurred during a pre-download storage check while SAF was enabled. One wonders how many more there could be.
+
+- [@Douile](https://github.com/Douile) found a way to minimize data usage for the reCAPTCHA challenge by loading an empty page, which prevents YouTube from staying loaded in the background and consuming data.
+
+- [@Douile](https://github.com/Douile) also fixed a small issue where one could press on the hidden Feed items while it was being refreshed, resulting in a stuck loading screen.
+
+- One last UI fix for this release, from [@Stypox](https://github.com/Stypox): The fix for the grid layout of channels in the Subscriptions tab accidentally made the layout less condensed, which has been reversed.
+
+- [@TiA4f8R](https://github.com/TiA4f8R) modified the regular expressions (or regexes, as the cool kids say it) used to find the dreaded cipher deobfuscation function. This fixes a 'Video not available' error for some music videos.
+
+## Nerd Talk
+
+- [@TacoTheDank](https://github.com/TacoTheDank) upgraded the Gradle version to 7.1.1 for both NewPipe and the Extractor.
+
+- [@gardenappl](https://github.com/gardenappl) fixed some typos in the comments for the Extractor code.
+
+- In case content is not available for some reason, [@Stypox](https://github.com/Stypox) made the app show the original error which caused it.
+
+- [@TacoTheDank](https://github.com/TacoTheDank) also upgraded some libraries:
+  
+  - AndroidX AppCompat 1.2.0 -> 1.3.1
+  - AndroidX Media 1.2.1 -> 1.3.1
+  - AndroidX Core 1.3.2 -> 1.6.0 
+  - AndroidX Fragment 1.3.5 -> 1.3.6 
+  - AndroidX Lifecycle 2.2.0 -> 2.3.1
+  - Markwon 4.6.0 -> 4.6.2
+  - Stetho 1.5.1 -> 1.6.0
+  - AutoService 1.0-rc7 -> 1.0
+
+- [@XiangRongLin](https://github.com/XiangRongLin) specified the `emulator-build` version of the Android Emulator in the GitHub Actions CI to fix a timeout error.
+
+- An earlier version fixed NewPipe so that it shut down properly instead of being killed instantly by the OS. [@XiangRongLin](https://github.com/XiangRongLin) cleaned up some related code as a follow-up, including a variable that was no longer needed. He also fixed some warnings emitted by Sonar.
+
+- [@XiangRongLin](https://github.com/XiangRongLin) also temporarily removed then re-added API level 29 to the Android Emulator in the CI, to avoid a timeout caused by the CI being stuck waiting for some user input to accept the Android SDK license. This was fixed upstream.
```

Comment:
IIRC, the first and third point were targeting the same problem.

Reply:
@TobiGr They're different looking to me:

> Timeout waiting for emulator to boot.

> sdkmanager is hanging the script because waits the user to accept the SDK license terms.

@XiangRongLin can confirm for us either way.

Reply:
They are related to each other, so they can be merged

The github action that we are using was broken causing the emulator to time out. Because of that, tests with API level 29 were commented out (third point). After the action was fixed and released, which they (not us) did by accepting the SDK license terms, our CI action got updated by specifying the `emulator-build` version (first point) and readding the API level 29 tests.

Reply:
Alright. I'll merge those points.

Reply:
```suggestion
- [@XiangRongLin](https://github.com/XiangRongLin) noticed that the API level 29 runner was timing out, so it was removed temporarily. The bug was fixed by specifying the `emulator-build` version of the Android Emulator in the GitHub Actions CI to fix the timeout error. It was caused by the runner waiting for some user input to accept the Android SDK license, and was fixed upstream.

- An earlier version fixed NewPipe so that it shut down properly instead of being killed instantly by the OS. [@XiangRongLin](https://github.com/XiangRongLin) cleaned up some related code as a follow-up, including a variable that was no longer needed. He also fixed some warnings emitted by Sonar.
```

Reply:
Is this fine?

Reply:
fine by me